presto-cache/src/main/java/com/facebook/presto/cache/alluxio/AlluxioCachingHdfsFileInputStream.java

```diff
@@ -15,36 +15,62 @@
 
 import alluxio.client.file.FileInStream;
 import alluxio.exception.ExceptionMessage;
+import org.apache.hadoop.fs.FSDataInputStream;
 import org.apache.hadoop.fs.PositionedReadable;
 import org.apache.hadoop.fs.Seekable;
 
 import java.io.EOFException;
 import java.io.IOException;
 import java.io.InputStream;
 
+import static com.google.common.base.Verify.verify;
+
 public class AlluxioCachingHdfsFileInputStream
         extends InputStream
         implements Seekable, PositionedReadable
 {
     private final FileInStream inputStream;
+    private final FSDataInputStream originInputStream;
+    private final boolean cacheValidationEnabled;
 
     public AlluxioCachingHdfsFileInputStream(FileInStream inputStream)
+    {
+        this(inputStream, null, false);
+    }
+
+    public AlluxioCachingHdfsFileInputStream(FileInStream inputStream,
+            FSDataInputStream originInputStream,
+            boolean cacheValidationEnabled)
     {
         this.inputStream = inputStream;
+        this.originInputStream = originInputStream;
+        this.cacheValidationEnabled = cacheValidationEnabled;
     }
 
     @Override
     public int read()
             throws IOException
     {
-        return inputStream.read();
+        int outByte = inputStream.read();
+        if (cacheValidationEnabled) {
+            verify(originInputStream.read() == outByte, "corrupted buffer at position " + getPos());
+        }
+        return outByte;
     }
 
     @Override
     public int read(long position, byte[] buffer, int offset, int length)
             throws IOException
     {
-        return inputStream.positionedRead(position, buffer, offset, length);
+        int bytesRead = inputStream.positionedRead(position, buffer, offset, length);
```

Comment:
`bytes`

Reply:
This is only reading one byte and the output is the byte, should it be bytes?